# Pandas Data Wrangling

### Homework

Before you begin, remember to import the necessary libraries.

In [2]:
import numpy as np
import pandas as pd

#### Standard Exercises

1. Load the `lifeexpectancy.csv` and the `drinks.csv` datasets from the `data` > `Life Expectancy` folder and save them, respectively, to the following two objects: `life_raw` and `drinks_raw`.

In [3]:
life_raw = pd.read_csv('data/Life Expectancy/lifeexpectancy.csv')
life_raw.sample(5)

GhoCode                                       GhoDisplay  \
3047  WHOSIS_000001                 Life expectancy at birth (years)   
4362  WHOSIS_000001                 Life expectancy at birth (years)   
1109  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
4432  WHOSIS_000001                 Life expectancy at birth (years)   
556   WHOSIS_000015                Life expectancy at age 60 (years)   

     PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
3047        PUBLISHED           Published      1990         1990        WPR   
4362        PUBLISHED           Published      2000         2000        AMR   
1109        PUBLISHED           Published      2000         2000        AFR   
4432        PUBLISHED           Published      1990         1990        WPR   
556         PUBLISHED           Published      2000         2000        EUR   

        RegionDisplay WorldBankIncomeGroupGroupCode  \
3047  Western Pacific                         WB_LI   
4362         Americas                        WB_LMI   
1109           Africa                         WB_LI   
4432  Western Pacific                        WB_LMI   
556            Europe                         WB_HI   

     WorldBankIncomeGroupDisplay CountryCode  \
3047                  Low_income         LAO   
4362         Lower_middle_income         PRY   
1109                  Low_income         TCD   
4432         Lower_middle_income         PHL   
556                  High_income         BEL   

                        CountryDisplay SexCode  SexDisplay  DisplayValue  \
3047  Lao People's Democratic Republic     MLE        Male            51   
4362                          Paraguay    BTSX  Both sexes            74   
1109                              Chad     MLE        Male            39   
4432                       Philippines    FMLE      Female            70   
556                            Belgium    FMLE      Female            24   

      Numeric  
3047     51.0  
4362     74.0  
1109     38.9  
4432     70.0  
556      24.0

In [4]:
drinks_raw = pd.read_csv('data/Life Expectancy/drinks.csv')
drinks_raw.sample(5)

country beer_servings spirit_servings wine_servings  \
9                Austria           279              75           191   
145  Sao Tome & Principe            56              38           140   
157          South Korea           140              16             9   
22              Botswana           173              35            35   
174                Tonga            36              21             5   

     total_litres_of_pure_alcohol  
9                             NaN  
145                           NaN  
157                           NaN  
22                            NaN  
174                           NaN

2. For each DataFrame, create a copy and save it to a new object (`life` and `drinks`), check its columns data type and make sure each column is in the expected format and whether there are any missing values. *(hint: remember the `.info()` method)* 

In [5]:
life = life_raw.copy()
life.head()

GhoCode                        GhoDisplay PublishStateCode  \
0  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
1  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
2  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
3  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
4  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   

  PublishStateDisplay  YearCode  YearDisplay RegionCode  \
0           Published      2013         2013        EMR   
1           Published      2013         2013        EMR   
2           Published      2013         2013        EMR   
3           Published      2012         2012        EMR   
4           Published      2012         2012        EMR   

           RegionDisplay WorldBankIncomeGroupGroupCode  \
0  Eastern Mediterranean                           NaN   
1  Eastern Mediterranean                           NaN   
2  Eastern Mediterranean                           NaN   
3  Eastern Mediterranean                         WB_LI   
4  Eastern Mediterranean                         WB_LI   

  WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  SexDisplay  \
0                         NaN         AFG    Afghanistan    BTSX  Both sexes   
1                         NaN         AFG    Afghanistan    FMLE      Female   
2                         NaN         AFG    Afghanistan     MLE        Male   
3                  Low_income         AFG    Afghanistan    BTSX  Both sexes   
4                  Low_income         AFG    Afghanistan    FMLE      Female   

   DisplayValue   Numeric  
0            61  61.25205  
1            62  61.97733  
2            61  60.51228  
3            60  60.00000  
4            61  61.00000

In [6]:
life.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6408 entries, 0 to 6407
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   GhoCode                        6408 non-null   object 
 1   GhoDisplay                     6408 non-null   object 
 2   PublishStateCode               6408 non-null   object 
 3   PublishStateDisplay            6408 non-null   object 
 4   YearCode                       6408 non-null   int64  
 5   YearDisplay                    6408 non-null   int64  
 6   RegionCode                     6408 non-null   object 
 7   RegionDisplay                  6408 non-null   object 
 8   WorldBankIncomeGroupGroupCode  5244 non-null   object 
 9   WorldBankIncomeGroupDisplay    5244 non-null   object 
 10  CountryCode                    6408 non-null   object 
 11  CountryDisplay                 6408 non-null   object 
 12  SexCode                        6408 non-null   o

In [7]:
drinks = drinks_raw.copy()
drinks.tail()

country beer_servings spirit_servings wine_servings  \
188  Venezuela           333             100             3   
189    Vietnam           111               2             1   
190      Yemen             6               0             0   
191     Zambia            32              19             4   
192   Zimbabwe            64              18             4   

     total_litres_of_pure_alcohol  
188                           NaN  
189                           NaN  
190                           NaN  
191                           NaN  
192                           NaN

In [8]:
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    object 
 2   spirit_servings               193 non-null    object 
 3   wine_servings                 193 non-null    object 
 4   total_litres_of_pure_alcohol  0 non-null      float64
dtypes: float64(1), object(4)
memory usage: 7.7+ KB


3. The `drinks` DataFrame has three columns (the ones ending with `_servings`) that should be interpreted as integer but are in fact stored as strings (ie: object). Find out why that is happening and replace any erroneous character with a zero. Then, convert each of those three columns to integers. *(hint: a quick way to check the contents of a column is to use the `.unique()` method)*

In [9]:
drinks['beer_servings'].unique()

array(['0', '89', '25', '245', '217', '102', '193', '21', '261', '279',
       '122', '42', '143', '142', '295', '263', '34', '23', '167', '76',
       '173', '31', '231', '88', '144', '57', '147', '240', '17', '15',
       '130', '79', '159', '1', '149', '37', '230', '93', '192', '361',
       '224', '52', '32', '162', '6', '92', '18', '20', '77', '127',
       '347', '8', '346', '133', '199', '53', '9', '28', '69', '234',
       '233', '5', '313', '63', '85', '82', '124', '58', '62', '281',
       '19', '343', '236', '?', '26', '13', '98', '238', '109', '12',
       '47', '376', '49', '251', '203', '78', '3', '188', '169', '22',
       '306', '285', '44', '213', '163', '71', '194', '297', '247', '43',
       '105', '56', '283', '157', '60', '196', '270', '225', '140', '284',
       '16', '171', '120', '128', '90', '152', '185', '2', '36', '99',
       '197', '51', '45', '206', '219', '115', '249', '333', '111', '64'],
      dtype=object)

In [10]:
drinks['spirit_servings'].unique()

array(['0', '132', '138', '57', '128', '25', '179', '72', '75', '46',
       '176', '63', '173', '373', '84', '114', '4', '41', '35', '145',
       '2', '252', '7', '56', '65', '1', '122', '124', '192', '76', '3',
       '254', '87', '137', '154', '170', '?', '44', '286', '147', '74',
       '69', '194', '133', '151', '98', '100', '117', '112', '438', '31',
       '302', '326', '215', '61', '118', '42', '97', '202', '21', '246',
       '22', '34', '216', '55', '29', '152', '244', '27', '15', '11',
       '68', '50', '226', '189', '6', '18', '88', '79', '5', '200', '71',
       '16', '104', '39', '160', '186', '67', '38', '131', '12', '293',
       '51', '157', '205', '315', '221', '13', '178', '60', '258', '156',
       '9', '237', '135', '126', '158', '101', '19'], dtype=object)

In [11]:
drinks['wine_servings'].unique()

array(['0', '54', '14', '312', '45', '221', '11', '212', '191', '5', '?',
       '7', '36', '42', '8', '13', '35', '16', '1', '94', '4', '100',
       '172', '3', '9', '74', '254', '113', '134', '278', '26', '2',
       '233', '59', '97', '370', '149', '175', '10', '218', '28', '21',
       '185', '78', '165', '237', '12', '6', '123', '62', '31', '56',
       '271', '86', '120', '18', '128', '190', '129', '23', '339', '167',
       '73', '24', '140', '127', '51', '116', '276', '81', '112', '32',
       '71', '186', '280', '19', '20', '195', '220', '84'], dtype=object)

In [12]:
def clean_function(col, find, replace, dtype): 
    return col.str.replace(find, replace, regex=False).astype(dtype)

In [13]:
for col in ['beer_servings', 'spirit_servings', 'wine_servings']:
    drinks[col] = clean_function(drinks[col], '?', '0', 'int32')

In [14]:
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int32  
 2   spirit_servings               193 non-null    int32  
 3   wine_servings                 193 non-null    int32  
 4   total_litres_of_pure_alcohol  0 non-null      float64
dtypes: float64(1), int32(3), object(1)
memory usage: 5.4+ KB


In [15]:
drinks['beer_servings'].unique()

array([  0,  89,  25, 245, 217, 102, 193,  21, 261, 279, 122,  42, 143,
       142, 295, 263,  34,  23, 167,  76, 173,  31, 231,  88, 144,  57,
       147, 240,  17,  15, 130,  79, 159,   1, 149,  37, 230,  93, 192,
       361, 224,  52,  32, 162,   6,  92,  18,  20,  77, 127, 347,   8,
       346, 133, 199,  53,   9,  28,  69, 234, 233,   5, 313,  63,  85,
        82, 124,  58,  62, 281,  19, 343, 236,  26,  13,  98, 238, 109,
        12,  47, 376,  49, 251, 203,  78,   3, 188, 169,  22, 306, 285,
        44, 213, 163,  71, 194, 297, 247,  43, 105,  56, 283, 157,  60,
       196, 270, 225, 140, 284,  16, 171, 120, 128,  90, 152, 185,   2,
        36,  99, 197,  51,  45, 206, 219, 115, 249, 333, 111,  64])

4. The `.drop_duplicates()` method allows you to remove duplicated rows from a DataFrame (read [here](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html) to find out more). Select the `CountryDisplay` and `RegionDisplay` columns from the `life` DataFrame and remove the duplicated rows. Save the final output in a new dataset called `country_region`. You just created a lookup-table that should have 195 rows and 2 columns. 

In [16]:
country_region = life[['CountryDisplay', 'RegionDisplay']].drop_duplicates()
print(country_region)
country_region.shape

                          CountryDisplay          RegionDisplay
0                            Afghanistan  Eastern Mediterranean
33                               Albania                 Europe
66                               Algeria                 Africa
99                               Andorra                 Europe
132                               Angola                 Africa
...                                  ...                    ...
6243  Venezuela (Bolivarian Republic of)               Americas
6276                            Viet Nam        Western Pacific
6309                               Yemen  Eastern Mediterranean
6342                              Zambia                 Africa
6375                            Zimbabwe                 Africa

[195 rows x 2 columns]


(195, 2)

5. Using the `.merge()` method, combine the and the `drinks` and the `country_region` datasets using a **left join** in order to **add the `RegionDisplay` field** to the `drinks` DataFrame. Call the new DataFrame object `drinks_region`.

In [17]:
drinks.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 193 entries, 0 to 192
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   country                       193 non-null    object 
 1   beer_servings                 193 non-null    int32  
 2   spirit_servings               193 non-null    int32  
 3   wine_servings                 193 non-null    int32  
 4   total_litres_of_pure_alcohol  0 non-null      float64
dtypes: float64(1), int32(3), object(1)
memory usage: 5.4+ KB


In [18]:
life.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6408 entries, 0 to 6407
Data columns (total 16 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   GhoCode                        6408 non-null   object 
 1   GhoDisplay                     6408 non-null   object 
 2   PublishStateCode               6408 non-null   object 
 3   PublishStateDisplay            6408 non-null   object 
 4   YearCode                       6408 non-null   int64  
 5   YearDisplay                    6408 non-null   int64  
 6   RegionCode                     6408 non-null   object 
 7   RegionDisplay                  6408 non-null   object 
 8   WorldBankIncomeGroupGroupCode  5244 non-null   object 
 9   WorldBankIncomeGroupDisplay    5244 non-null   object 
 10  CountryCode                    6408 non-null   object 
 11  CountryDisplay                 6408 non-null   object 
 12  SexCode                        6408 non-null   o

In [19]:
drinks_region = drinks.merge(country_region, how='left', left_on='country', right_on='CountryDisplay')
drinks_region.sample(5)

country  beer_servings  spirit_servings  wine_servings  \
142   Rwanda             43                2              0   
72     Haiti              1              326              1   
176  Tunisia             51                3             20   
58      Fiji             77               35              1   
124    Niger              3                2              1   

     total_litres_of_pure_alcohol CountryDisplay          RegionDisplay  
142                           NaN         Rwanda                 Africa  
72                            NaN          Haiti               Americas  
176                           NaN        Tunisia  Eastern Mediterranean  
58                            NaN           Fiji        Western Pacific  
124                           NaN          Niger                 Africa

6. Rename the `RegionDisplay` column to `region` and reorder the columns in the `drinks_region` DataFrame so to have the following sequence: ['region', 'country', 'beer_servings', 'spirit_servings', 'wine_servings'].

In [20]:
drinks_region = drinks_region.rename(columns={'RegionDisplay':'region'})
drinks_region

country  beer_servings  spirit_servings  wine_servings  \
0    Afghanistan              0                0              0   
1        Albania             89              132             54   
2        Algeria             25                0             14   
3        Andorra            245              138            312   
4         Angola            217               57             45   
..           ...            ...              ...            ...   
188    Venezuela            333              100              3   
189      Vietnam            111                2              1   
190        Yemen              6                0              0   
191       Zambia             32               19              4   
192     Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol CountryDisplay                 region  
0                             NaN    Afghanistan  Eastern Mediterranean  
1                             NaN        Albania                 Europe  
2                             NaN        Algeria                 Africa  
3                             NaN        Andorra                 Europe  
4                             NaN         Angola                 Africa  
..                            ...            ...                    ...  
188                           NaN            NaN                    NaN  
189                           NaN            NaN                    NaN  
190                           NaN          Yemen  Eastern Mediterranean  
191                           NaN         Zambia                 Africa  
192                           NaN       Zimbabwe                 Africa  

[193 rows x 7 columns]

In [21]:
drinks_region = drinks_region[['region', 'country', 'spirit_servings', 'wine_servings']]
drinks_region

region      country  spirit_servings  wine_servings
0    Eastern Mediterranean  Afghanistan                0              0
1                   Europe      Albania              132             54
2                   Africa      Algeria                0             14
3                   Europe      Andorra              138            312
4                   Africa       Angola               57             45
..                     ...          ...              ...            ...
188                    NaN    Venezuela              100              3
189                    NaN      Vietnam                2              1
190  Eastern Mediterranean        Yemen                0              0
191                 Africa       Zambia               19              4
192                 Africa     Zimbabwe               18              4

[193 rows x 4 columns]

7. Find out which `region` has the **highest average spirit_servings**.

In [22]:
drinks_region.groupby('region', as_index=False)['spirit_servings'].mean().sort_values('spirit_servings', ascending=False)

region  spirit_servings
1               Americas       143.148148
3                 Europe       128.142857
5        Western Pacific        78.863636
4        South_East Asia        53.777778
2  Eastern Mediterranean        21.578947
0                 Africa        17.487805

#### Advanced Exercise

1. Using the `.merge()` method, combine the `life` and the `drinks` datasets using a **left join** on the appropriate field. Save the new dataset in a new DataFrame object called `life_drinks_raw`.

In [34]:
life_drinks_raw = life.merge(drinks, how='left', left_on='CountryDisplay', right_on='country')
life_drinks_raw.head()

GhoCode                        GhoDisplay PublishStateCode  \
0  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
1  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
2  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
3  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
4  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   

  PublishStateDisplay  YearCode  YearDisplay RegionCode  \
0           Published      2013         2013        EMR   
1           Published      2013         2013        EMR   
2           Published      2013         2013        EMR   
3           Published      2012         2012        EMR   
4           Published      2012         2012        EMR   

           RegionDisplay WorldBankIncomeGroupGroupCode  \
0  Eastern Mediterranean                           NaN   
1  Eastern Mediterranean                           NaN   
2  Eastern Mediterranean                           NaN   
3  Eastern Mediterranean                         WB_LI   
4  Eastern Mediterranean                         WB_LI   

  WorldBankIncomeGroupDisplay  ... CountryDisplay SexCode  SexDisplay  \
0                         NaN  ...    Afghanistan    BTSX  Both sexes   
1                         NaN  ...    Afghanistan    FMLE      Female   
2                         NaN  ...    Afghanistan     MLE        Male   
3                  Low_income  ...    Afghanistan    BTSX  Both sexes   
4                  Low_income  ...    Afghanistan    FMLE      Female   

  DisplayValue   Numeric      country beer_servings  spirit_servings  \
0           61  61.25205  Afghanistan           0.0              0.0   
1           62  61.97733  Afghanistan           0.0              0.0   
2           61  60.51228  Afghanistan           0.0              0.0   
3           60  60.00000  Afghanistan           0.0              0.0   
4           61  61.00000  Afghanistan           0.0              0.0   

   wine_servings  total_litres_of_pure_alcohol  
0            0.0                           NaN  
1            0.0                           NaN  
2            0.0                           NaN  
3            0.0                           NaN  
4            0.0                           NaN  

[5 rows x 21 columns]

2. Create a copy of `life_drinks_raw` called `life_drinks` where you filter the above DataFrame using the following conditions (check out the solution on [this Stack Overflow page](https://stackoverflow.com/questions/22546425/how-to-implement-a-boolean-search-with-multiple-columns-in-pandas) to get a cue on how to filter data on multiple columns in a pandas DataFrame): 

- `GhoDisplay` = 'Life expectancy at birth (years)'
- `WorldBankIncomeGroupDisplay` = null *(hint: remember the isnull() method)*
- `SexDisplay` != 'Both sexes'

In [35]:
life_drinks = life_drinks_raw[(life_drinks_raw['GhoDisplay']=='Life expectancy at birth (years)') &
                              (life_drinks_raw['WorldBankIncomeGroupDisplay'].isnull()) &
                              (life_drinks_raw['SexDisplay']!='Both sexes')]

In [36]:
life_drinks.head()

GhoCode                        GhoDisplay PublishStateCode  \
1   WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
2   WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
34  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
35  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   
67  WHOSIS_000001  Life expectancy at birth (years)        PUBLISHED   

   PublishStateDisplay  YearCode  YearDisplay RegionCode  \
1            Published      2013         2013        EMR   
2            Published      2013         2013        EMR   
34           Published      2013         2013        EUR   
35           Published      2013         2013        EUR   
67           Published      2013         2013        AFR   

            RegionDisplay WorldBankIncomeGroupGroupCode  \
1   Eastern Mediterranean                           NaN   
2   Eastern Mediterranean                           NaN   
34                 Europe                           NaN   
35                 Europe                           NaN   
67                 Africa                           NaN   

   WorldBankIncomeGroupDisplay  ... CountryDisplay SexCode SexDisplay  \
1                          NaN  ...    Afghanistan    FMLE     Female   
2                          NaN  ...    Afghanistan     MLE       Male   
34                         NaN  ...        Albania    FMLE     Female   
35                         NaN  ...        Albania     MLE       Male   
67                         NaN  ...        Algeria    FMLE     Female   

   DisplayValue   Numeric      country beer_servings  spirit_servings  \
1            62  61.97733  Afghanistan           0.0              0.0   
2            61  60.51228  Afghanistan           0.0              0.0   
34           76  75.50800      Albania          89.0            132.0   
35           73  72.82101      Albania          89.0            132.0   
67           74  73.51464      Algeria          25.0              0.0   

    wine_servings  total_litres_of_pure_alcohol  
1             0.0                           NaN  
2             0.0                           NaN  
34           54.0                           NaN  
35           54.0                           NaN  
67           14.0                           NaN  

[5 rows x 21 columns]

3. Reorder the columns in the `life_drinks` DataFrame so to have the following sequence: 
['GhoDisplay', 'YearDisplay', 'YearDisplay', 'RegionDisplay', 'CountryDisplay', 'SexDisplay', 'Numeric', 'beer_servings',
       'spirit_servings', 'wine_servings']

In [37]:
life_drinks = life_drinks[['GhoDisplay', 'YearDisplay', 'YearDisplay', 'RegionDisplay', 'CountryDisplay', 'SexDisplay', 'Numeric', 'beer_servings',
       'spirit_servings', 'wine_servings']]
life_drinks.head()

GhoDisplay  YearDisplay  YearDisplay  \
1   Life expectancy at birth (years)         2013         2013   
2   Life expectancy at birth (years)         2013         2013   
34  Life expectancy at birth (years)         2013         2013   
35  Life expectancy at birth (years)         2013         2013   
67  Life expectancy at birth (years)         2013         2013   

            RegionDisplay CountryDisplay SexDisplay   Numeric  beer_servings  \
1   Eastern Mediterranean    Afghanistan     Female  61.97733            0.0   
2   Eastern Mediterranean    Afghanistan       Male  60.51228            0.0   
34                 Europe        Albania     Female  75.50800           89.0   
35                 Europe        Albania       Male  72.82101           89.0   
67                 Africa        Algeria     Female  73.51464           25.0   

    spirit_servings  wine_servings  
1               0.0            0.0  
2               0.0            0.0  
34            132.0           54.0  
35            132.0           54.0  
67              0.0           14.0

4. Rename the following columns as specified: 

- Numeric --> LifeExpectancy
- beer_servings = BeerServings
- spirit_servings = SpiritServings
- wine_servings = WineServings

In [38]:
life_drinks = life_drinks.rename(columns={'Numeric':'LifeExpectancy', 
                                          'beer_servings':'BeerServings', 
                                          'spirit_servings':'SpiritServings', 
                                          'wine_servings':'WineServings'})
life_drinks.head()

GhoDisplay  YearDisplay  YearDisplay  \
1   Life expectancy at birth (years)         2013         2013   
2   Life expectancy at birth (years)         2013         2013   
34  Life expectancy at birth (years)         2013         2013   
35  Life expectancy at birth (years)         2013         2013   
67  Life expectancy at birth (years)         2013         2013   

            RegionDisplay CountryDisplay SexDisplay  LifeExpectancy  \
1   Eastern Mediterranean    Afghanistan     Female        61.97733   
2   Eastern Mediterranean    Afghanistan       Male        60.51228   
34                 Europe        Albania     Female        75.50800   
35                 Europe        Albania       Male        72.82101   
67                 Africa        Algeria     Female        73.51464   

    BeerServings  SpiritServings  WineServings  
1            0.0             0.0           0.0  
2            0.0             0.0           0.0  
34          89.0           132.0          54.0  
35          89.0           132.0          54.0  
67          25.0             0.0          14.0

5. Create a new DataFrame, call it `life_drinks_13`, where you show the average `LifeExpectancy` and `SpiritServings` by `RegionDisplay` in 2013. Can you see a relationship between the two metrics? If so, would you trust this data to make any broad conclusions?

*(Hint: remember to include the `as_index=False` parameter in the groupby so to aviod multi-index DataFrames)*

In [41]:
life_drinks_13 = life_drinks[life_drinks['YearDisplay']==2013].groupby(by=['RegionDisplay'], as_index=False)[['LifeExpectancy','SpiritServings']].mean()

ValueError: cannot reindex on an axis with duplicate labels

6. Now create a new DataFrame, call it `life_exp_13`, where you show the average `LifeExpectancy` by `RegionDisplay` and `SexDisplay` in 2013.

7. Starting from the `life_exp_13` DataFrame and using an **inner join**, create a DataFrame named `life_exp_13_gndr` that shows one row per `RegionDisplay` (6 rows) and two columns, each containing the average male and female life expectancy (relative to each region). 

    - Then rename and reorder the DataFrame's columns so to end up with the following column names: ['RegionDisplay', 'LifeExpMale', 'LifeExpFemale']. 
    - Which `DisplayRegion` has the highest gap in years between males and females?